modules/ethereum/src/lib.rs

```diff
@@ -251,9 +285,25 @@ pub trait Trait: frame_system::Trait {
 
 decl_module! {
 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
-		/// Import Aura chain headers. Ignores non-fatal errors (like when known
-		/// header is provided), rewards for successful headers import and penalizes
-		/// for fatal errors.
+		/// Import single Aura header. Requires transaction to be **UNSIGNED**.
+		pub fn unsigned_import_header(origin, header: Header, receipts: Option<Vec<Receipt>>) {
```

Comment:
Why don't we have a function to import a chain of unsigned headers just like we do with signed headers? It seems like weird to me that the APIs don't totally match up

Reply:
One of good things that we're getting with unsigned transactions is that they're auto-deduplicated by the txpool/network. I.e. now unsigned transaction is just a `Call(unsigned_import_header(Header, Option<Vec<Receipt>>)).encode` without any additional data. So if we have that transaction in the pool AND someone sends us the same transaction again, we'll see that the hash is the same => won't bother with checks, etc. If we would allow to submit several header at once (without any strict rules on how many/when/...), then we'll get a case when actor#1 which submit tx1 with headers (101, 102, 103), actor#2 which submits tx2 with headers (101, 102), actor#3 which submits tx3 with headers (101). All these transactions have different hashes && `validate_unsigned()` would be called. Then we have a problems with `validate_unsigned()` - which transactions to accept, and which to reject - if we'll accept all transactions, then we would allow to spam blocks without any payment (header 101 is provided by every transaction). If we'll accept single transaction, then we may miss these additional headers.

That said, having multiple headers in single unsigned transaction is possible if we'll enforce some rules that will reintroduce deduplication - like we only accept 16 headers (starting with number N so that N % 16 == 0) in single unsigned transaction when we know we are behind the chain tip and only 1 header otherwise. But we won't have the same API anyway. And I have a feeling that this optimization will reveal another bunch of problems that we'll need to solve. So let's not introduce it here for now.

While writing this, I've realized that I'm not checking if `receipts` are provided (when required) in `validate_unsigned()`. And I need to, because otherwise: (1) we can have two transactions for the same header with different hashes (2) transaction that is missing required `receipts` could push good transaction out of pool (2) this allow submitter to spam blocks with unneeded data. So please do not resolve this comment - I'll fix it

Reply:
OK - so in addition to verifying that receipts are provided when required, we also need to verify that receipts are valid. Hopefully, this wouldn't happen too often - normally it only comes with validators set change. But not sure how often we'll see false positives.